raven/src/main/java/org/opencadc/raven/ProtocolsGenerator.java

```diff
@@ -341,4 +361,62 @@ private boolean isAvailable(URI resourceID) {
         return true;
     }
 
+    /**
+     * Compare two StorageSite's, giving higher ranking to the site with a
+     * namespace matching an Artifact URI.
+     */
+    static class PrioritizingStorageSiteComparator implements Comparator<StorageSite> {
+
+        private final Map<URI, StorageSiteRule> siteRules;
+        private final URI artifactURI;
+        private InetAddress clientIP;
+
+        public PrioritizingStorageSiteComparator(Map<URI, StorageSiteRule> siteRules,
+                                                 URI artifactURI, InetAddress clientIP) {
+            this.siteRules = siteRules;
+            this.artifactURI = artifactURI;
+            this.clientIP = clientIP;
+        }
+
+        @Override
+        public int compare(StorageSite site1, StorageSite site2) {
+
+            boolean site1Match = hasMatchingNamespace(site1);
+            boolean site2match = hasMatchingNamespace(site2);
+
```

Comment:
First: the logic in the `hasMatchingNamespace` is wrong and that method is useless.

What you want here is something like:
```
// get namespace for each site
Namespace ns1 = siteRules.get(site1);
Namespace ns2 = siteRules.get(site2);

boolean site1Match = ns1.matches(artifactURI);
boolean site2Match = ns2.matches(artifactURI);
```
Above is pseudo-code and doesn't handle when ns1 or ns2 are null, which is possible when there are no rules for a site so it still needs some work. However, while reviewing Namespace it had occurred to me you would want a method like
```
public boolean matches(URI artifactURI)
```
and it is because I expected the above kind of usage.


Second: for the if/else if/else below, it is better to write it as:
```
if (site1Match && !site2match) {
     return 1;
}
if (!site1Match && site2match) {
     return -1;
}
// default: order of site names??
return site1.getResourceID().compateTo(site2.getResourceID());
```

Reply:
Added the `matches` method to Namespace, and rewrote the PrioritizingStorageSiteComparator compare to use the `matches` method.

Reply:
Comparator still does order StorageSite instances correctly. Read the documentation and pay attention to the concept of "consistent with equals":. Specifically, if there is no StorageSiteRule for a two sites, that does not make them equal (line 401). It has to fall through to the default order. 

You clearly do not have unit tests that cover this scenario; if you did, one of those "equal" sites would not be added to the set and one URL would be missing from the resulting prioritised list of URLs.

Reply:
Missed that scenario. Updated the comparator to use the default StorageSite ordering when it is unable to determine the ordering between sites. I added a separate unit test for the the comparator class to make sure it is doing the right thing.